langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/BedrockAnthropicMessageChatModel.java

```diff
@@ -53,47 +59,164 @@ protected Map<String, Object> getRequestParameters(String prompt) {
         parameters.put("anthropic_version", anthropicVersion);
         return parameters;
     }
-    
+
     @Override
     public Response<AiMessage> generate(List<ChatMessage> messages) {
+        return generate(messages, (ToolSpecification) null);
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, ToolSpecification toolSpecification) {
+        return generate(
+                messages,
+                Optional.ofNullable(toolSpecification)
+                        .map(Collections::singletonList)
+                        .orElse(Collections.emptyList())
+        );
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
         final String system = getAnthropicSystemPrompt(messages);
-        
+
         List<BedrockAnthropicMessage> formattedMessages = getAnthropicMessages(messages);
-        
+
         Map<String, Object> parameters = getRequestParameters(null);
         parameters.put("messages", formattedMessages);
         parameters.put("system", system);
-        
+
+        if (!toolSpecifications.isEmpty()) {
+            validateModelIdWithToolsSupport();
+            parameters.put("tools", toAnthropicToolSpecifications(toolSpecifications));
+        }
+
         final String body = Json.toJson(parameters);
-        
+
         InvokeModelResponse invokeModelResponse = withRetry(() -> invoke(body), getMaxRetries());
         final String response = invokeModelResponse.body().asUtf8String();
         BedrockAnthropicMessageChatModelResponse result = Json.fromJson(response, getResponseClassType());
-        
-        return new Response<>(new AiMessage(result.getOutputText()),
+
+        return new Response<>(
+            aiMessageFrom(result),
             result.getTokenUsage(),
-            result.getFinishReason());
+            result.getFinishReason()
+        );
     }
-    
+
+    private void validateModelIdWithToolsSupport() {
```

Comment:
can't we delegate this validation to Bedrock?

Reply:
Yes. But Bedrock returns something like an invalid request. My goal with this method was to make it explicit which versions of Anthropic supports the tools.